pytradfri/aiocoap_api.py

```diff
@@ -0,0 +1,88 @@
+"""Coap implementation using aiocoap."""
+import json
+import logging
+
+import asyncio
+from aiocoap import Message, Context, GET, PUT
+from aiocoap.transports.tinydtls import PSK_STORE
+from pytradfri.command import Command
+from .parser import process_output
+
+_LOGGER = logging.getLogger(__name__)
+
+
+def api_factory(host, security_code):
+    """Generate a request method."""
+
+    PSK_STORE[b'Client_identity'] = security_code.encode('utf-8')
+
+    @asyncio.coroutine
+    def request(api_command):
+        """Make a request."""
+        method = api_command.method
+        path = api_command.path
+        data = api_command.data
+        parse_json = api_command.parse_json
+        timeout = api_command.timeout
+        url = api_command.url(host)
+        callback = api_command.callback
+
+        if api_command.observe:
+            yield from _observe(url, callback, api_command.observe_duration)
+            return
+
+        kwargs = {}
+
+        if data is not None:
+            kwargs['payload'] = json.dumps(data).encode('utf-8')
+            _LOGGER.debug('Executing %s %s %s: %s', host, method, path, data)
+        else:
+            _LOGGER.debug('Executing %s %s %s', host, method, path)
+
+        api_method = GET
+        if method == 'put':
+            api_method = PUT
+
+        protocol = yield from Context.create_client_context()
+        msg = Message(code=api_method, uri=url, **kwargs)
+        res = yield from protocol.request(msg).response
+        _LOGGER.debug("RECEIVED STATUS", res.code)
+        _LOGGER.debug("RECEIVED PAYLOAD", res.payload.decode('utf-8'))
+
+        api_command.result = process_output(res, parse_json)
```

Comment:
I took another look at `process_output` and the only similarity will likely end up being parsing the JSON, unless I've missed something, it looks like the aiocoap representation of codes is different?

Raising exceptions in callbacks always feels messy to me. We could lazily process the response, raising an exception when the result is requested? Although, I can't visualise this using the command pattern for requests such as `PUT`.